### ISSUE TITLE
Sitemap generation

### ISSUE THREAD

**Issue by kitswas** [doxygen/doxygen#10023]
This is an enhancement/feature request.

HTML pages generated via Doxygen have a **lot** of 'dynamic' content.  
Please see [this documentation site, for example,](https://kitswas.github.io/CPU-scheduling-simulator/) with JS enabled and disabled to feel the difference. Without JS, only the primary content area of the `index.html` page is shown.  
Web crawlers do not execute JS scripts. So, the only way to inform them without using a sitemap is to submit the links manually to each search engine. _(Ouch)_

Could we have Doxygen create a sitemap when HTML output is enabled?

Here's the link to the [XML schema for the Sitemap protocol](https://www.sitemaps.org/protocol.html).

**Comment by kitswas** [comment]
And since the JS is not executed, the links are not present in the HTML.  
Hence, the search engines cannot follow them to build up a network of links either.

**Comment by kitswas** [comment]
Here's a sample sitemap created using [a site on the internet](https://www.xml-sitemaps.com/) for the site used as an example above.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<urlset
      xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9
            http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd">
<!-- created with Free Online Sitemap Generator www.xml-sitemaps.com -->


<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>1.00</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classSimulator.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.80</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classFCFS.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.80</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classRR.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.80</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classScheduler.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.80</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classCFS.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.80</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classSimulator-members.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.64</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classProcess.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.64</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/process_8hpp.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.64</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/simulator_8hpp.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.64</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/simulator_8cpp.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.64</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classFCFS-members.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.64</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/graph_legend.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.64</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classLogger.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.64</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/scheduler_8hpp.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.64</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/FCFS_8cpp.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.64</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classRR-members.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.64</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/RR_8cpp.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.64</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classScheduler-members.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.64</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classCFS-members.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.64</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/CFS_8cpp.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.64</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classProcess-members.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.51</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/process_8cpp.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.51</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/dir_68267d1309a1af8e8297ef4c3efbcdba.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.51</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/printer_8hpp.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.51</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classLogger-members.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.51</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/logger_8hpp.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.51</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/min__heap_8hpp.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.51</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/rbtree_8hpp.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.51</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/structRRProcess.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.51</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/structCFSProcess.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.51</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/main_8cpp.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.41</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/min__heap_8cpp.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.41</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/printer_8cpp.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.41</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/process__creator_8cpp.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.41</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/process__creator_8hpp.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.41</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/rbtree_8cpp.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.41</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classLog.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.41</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classMinHeap.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.41</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classRBTree.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.41</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/structRRProcess-members.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.41</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/structCFSProcess-members.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.41</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classProcessCreator.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.33</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classLog-members.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.33</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classMinHeap-members.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.33</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classRBTree-members.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.33</priority>
</url>
<url>
  <loc>https://kitswas.github.io/CPU-scheduling-simulator/classProcessCreator-members.html</loc>
  <lastmod>2023-04-22T06:40:20+00:00</lastmod>
  <priority>0.26</priority>
</url>

</urlset>
```

**Comment by kitswas** [comment]
The problem with generating sitemaps separately is that you have to do it every time there's a change in your website.  
If Doxygen did it while creating documentation, the entire process would be automated.  
Once a sitemap is discovered by crawlers, it is scanned periodically for changes.  
Only updated/new links are indexed. This reduces the amount of load on your hosting servers.

**Comment by albert-github** [comment]
> The problem with generating sitemaps separately is that you have to do it every time there's a change in your website.
> If Doxygen did it while creating documentation, the entire process would be automated.
> We could just set it up and forget about it.

so you could do it in your, automatic, deployment process, i.e. run doxygen followed by the process running the generation of the site map.

In the https://www.sitemaps.org/protocol.htmlI read  (bold is mine):
> `<loc>`  required  URL of the page. This URL **must begin with the protocol (such as http)** and end with a trailing slash, if your web server requires it. This value must be less than 2,048 characters.

which means that you have to know beforehand where the results will be deployed.
This again would be better of in a separate deployment process generating the site map as here the deployment destination is known and could be given as an argument to this process. (This can also be done by doxygen by specifying it as part of a new setting like `GENERATE_SITEMAP=<deployment_path>`).

**Comment by kitswas** [comment]
> In the https://www.sitemaps.org/protocol.htmlI read  (bold is mine):
> > `<loc>`  required  URL of the page. This URL **must begin with the protocol (such as http)** and end with a trailing slash, if your web server requires it. This value must be less than 2,048 characters.
> 
> which means that you have to know beforehand where the results will be deployed.
> This again would be better of in a separate deployment process generating the site map as here the deployment destination is known and could be given as an argument to this process. (This can also be done by doxygen by specifying it as part of a new setting like `GENERATE_SITEMAP=<deployment_path>`).
> 

The new setting idea was exactly what I was thinking of.

We could also have Doxygen write the timestamp into the `<lastmod>` parameter.

> > The problem with generating sitemaps separately is that you have to do it every time there's a change in your website.
> > If Doxygen did it while creating documentation, the entire process would be automated.
> > We could just set it up and forget about it.
> 
> so you could do it in your, automatic, deployment process, i.e. run doxygen followed by the process running the generation of the site map.

Fair point.  
However, only Doxygen knows whether all the HTML files produced are supposed to be displayed to a visitor. For instance, some of them could be error pages.  
Now we wouldn't want a 404 page to be shown in a Google search, would we?

**Comment by albert-github** [comment]
All pages that doxygen generates should be valid pages and there are, as far as I know, no 404 pages (and if they are present this constitutes a bug that should be fixed).Doxygen does know which files it generates, but these are / should be the only ones that are on the output directories.

**Comment by albert-github** [comment]
> We could also have Doxygen write the timestamp into the `<lastmod>` parameter.

I don't think this doesn't make much sense as each time doxygen generates the documentation all HTML files are newly generated and thus would get a new `<lastmod>` data.

**Comment by kitswas** [comment]
> each time doxygen generates the documentation all HTML files are newly generated

I was unaware. Thank you for letting me know.

**Comment by albert-github** [comment]
I've just pushed a proposed patch, pull request #10029

**Comment by albert-github** [comment]
Code has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.7.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
